src/cardano/CardanoTools.ts

```diff
@@ -0,0 +1,191 @@
+import * as Cardano from '@emurgo/cardano-serialization-lib-nodejs'
+import { div } from 'biggystring'
+import { entropyToMnemonic, mnemonicToEntropy, validateMnemonic } from 'bip39'
+import {
+  EdgeCurrencyInfo,
+  EdgeCurrencyTools,
+  EdgeEncodeUri,
+  EdgeIo,
+  EdgeMetaToken,
+  EdgeParsedUri,
+  EdgeWalletInfo
+} from 'edge-core-js/types'
+import { base16 } from 'rfc4648'
+
+import { PluginEnvironment } from '../common/innerPlugin'
+import { encodeUriCommon, parseUriCommon } from '../common/uriHelpers'
+import { getLegacyDenomination } from '../common/utils'
+import {
+  asCardanoPrivateKeys,
+  asSafeCardanoWalletInfo,
+  CardanoNetworkInfo,
+  EpochParams,
+  SafeCardanoWalletInfo
+} from './cardanoTypes'
+
+export class CardanoTools implements EdgeCurrencyTools {
+  io: EdgeIo
+  currencyInfo: EdgeCurrencyInfo
+  networkInfo: CardanoNetworkInfo
+  epochParams?: EpochParams
+
+  constructor(env: PluginEnvironment<CardanoNetworkInfo>) {
+    const { currencyInfo, io, networkInfo } = env
+    this.currencyInfo = currencyInfo
+    this.io = io
+    this.networkInfo = networkInfo
+  }
+
+  derivePrivateKeys(mnemonic: string): {
+    accountKey: Cardano.Bip32PrivateKey
+    privateKey: Cardano.Bip32PrivateKey
+  } {
+    const mnemonicEntropy = mnemonicToEntropy(mnemonic)
+    const privateKey = Cardano.Bip32PrivateKey.from_bip39_entropy(
+      base16.parse(mnemonicEntropy),
+      new Uint8Array()
+    )
+
+    function harden(num: number): number {
+      return 0x80000000 + num
+    }
+
+    const accountKey = privateKey
+      .derive(harden(1852)) // purpose
+      .derive(harden(1815)) // coin type
+      .derive(harden(0)) // account #0
+
+    return {
+      accountKey,
+      privateKey
+    }
+  }
+
+  async getDisplayPrivateKey(
+    privateWalletInfo: EdgeWalletInfo
+  ): Promise<string> {
+    const { pluginId } = this.currencyInfo
+    const keys = asCardanoPrivateKeys(pluginId)(privateWalletInfo.keys)
+    return keys.mnemonic
+  }
+
+  async getDisplayPublicKey(publicWalletInfo: EdgeWalletInfo): Promise<string> {
+    const { keys } = asSafeCardanoWalletInfo(publicWalletInfo)
+    return keys.publicKey
+  }
+
+  async importPrivateKey(mnemonic: string): Promise<Object> {
+    const { pluginId } = this.currencyInfo
+    const isValid = validateMnemonic(mnemonic)
+    if (!isValid) throw new Error('Invalid mnemonic')
+    const mnemonicEntropy = mnemonicToEntropy(mnemonic)
+
+    // Validate input
+    Cardano.Bip32PrivateKey.from_bip39_entropy(
+      base16.parse(mnemonicEntropy),
+      new Uint8Array()
+    )
+
+    return { [`${pluginId}Mnemonic`]: mnemonic }
+  }
+
+  async createPrivateKey(walletType: string): Promise<Object> {
+    const type = walletType.replace('wallet:', '')
```

Comment:
Just put this check at the top, and then you can get rid of the string parsing and the extra indentation level:

```ts
if (walletType !== this.currencyInfo.walletType) {
 throw new Error('InvalidWalletType')
}
```